### PR TITLE
Git subdir

### PR DESCRIPTION
**Pull request by Fuuzetsu**

### CODE REVIEW

**Comment by jkarni** [comment]
LGTM - the only minor quibble is that maybe a new datatype would be clearer/safer than `(String, Maybe String)`

File: hpack.cabal [hunk 1]
```diff
@@ -31,6 +31,7 @@ executable hpack
     , deepseq
     , directory
     , filepath
+    , split
```
Comment (sol) [inline hpack.cabal]: I think `text` also provides the functionality we need from `split` and is already a transitive dependency, so I would prefer to use that.

File: test/RunSpec.hs [hunk 2]
```diff
@@ -102,6 +102,19 @@ spec = do
...
+          "name: foo"
+        , "version: 0.0.0"
+        , "build-type: Simple"
+        , "cabal-version: >= 1.10"
+        , ""
+        , "source-repository head"
+        , "  type: git"
+        , "  location: https://github.com/hspec/hspec"
+        , "  subdir: driver"
+        ]
```
Comment (sol) [inline test/RunSpec.hs]: Do you think we could test `renderSourceRepository` independently?  I think the current testing approach does not scale.

File: test/ConfigSpec.hs [hunk 3]
```diff
@@ -177,7 +177,16 @@ spec = around_ (inTempDirectoryNamed "foo") $ do
...
+      packageSourceRepository c `shouldBe`
+        Just ("https://github.com/hspec/hspec", Nothing)
+
+    it "accepts third part of github URL as subdir" $ do
+      writeFile "package.yaml" [i|
+        github: hspec/hspec/driver
+        |]
+      Right (_, c) <- readPackageConfig "package.yaml"
+      packageSourceRepository c `shouldBe`
+        Just ("https://github.com/hspec/hspec", Just "driver")
```
Comment (sol) [inline test/ConfigSpec.hs]: If you change `driver` to `hspec-core` the example is more realistic ;).

File: src/Config.hs [hunk 4]
```diff
@@ -159,7 +162,7 @@ data Package = Package {
 , packageLicense :: Maybe String
 , packageLicenseFile :: Maybe FilePath
 , packageExtraSourceFiles :: [FilePath]
-, packageSourceRepository :: Maybe String
+, packageSourceRepository :: Maybe (String, Maybe String)
```
Comment (sol) [inline src/Config.hs]: I think at this point I'm also in favor of adding a proper `SourceRepository` type (kudos to @jkarni).

**Comment by Fuuzetsu** [comment]
Addressed the comments in new commits. Once everything is good I'll squash it of course.

Regarding `splitOn`, I replaced it with `Data.Text.split` as advised. There were multiple ways to do it: 

1. parse out Text from the yaml file, split what we get out and from there just use `String` so everything else works
2. Do the above but without converting to String. This made it a bit ugly as we were now mixing String and Text here and there.
3. Parse out as `String` but convert to `Text` just for the split.

I opted for 1. Option 3 would work too and would keep the `PackageConfig` field type unchanged if that's desired, very simple change. Option 2 could be nicer if we went ahead and refactored everything to just use `Text`.

File: src/Config.hs [hunk 5]
```diff
@@ -186,6 +188,16 @@ data Executable = Executable {
 , executableCppOptions :: [CppOption]
 } deriving (Eq, Show)
 
+data GithubConfig = GithubConfig {
+  githubConfigUrl :: String
+, githubConfigSubdir :: Maybe String
+} deriving (Eq, Show)
```
Comment (sol) [inline src/Config.hs]: Can we rename this type into `SourceRepository`?

File: src/Config.hs [hunk 6]
```diff
@@ -261,21 +273,28 @@ mkPackage (CaptureUnknownFields unknownFields PackageConfig{..}) = do
       where
         f name = "Specified source-dir " ++ show name ++ " does not exist"
 
-    github = ("https://github.com/" ++) <$> packageConfigGithub
+    github :: Maybe GithubConfig
+    github = modifyGithubUrl ("https://github.com/" <>) <$>
+             maybe Nothing (Just . splitGithub) packageConfigGithub
```
Comment (sol) [inline src/Config.hs]: ^ please don't align

File: src/Config.hs [hunk 7]
```diff
@@ -261,21 +273,28 @@ mkPackage (CaptureUnknownFields unknownFields PackageConfig{..}) = do
...
+      where
+        splitGithub str = case T.unpack <$> T.split (== '/') str of
```
Comment (sol) [inline src/Config.hs]: Using `fmap` on lists I consider code obfuscation.  See also https://ro-che.info/articles/2015-02-12-foldable-traversable-parametricity by @feuerbach.

File: src/Config.hs [hunk 8]
```diff
@@ -261,21 +273,28 @@ mkPackage (CaptureUnknownFields unknownFields PackageConfig{..}) = do
...
+            GithubConfig (user <> "/" <> repo) (Just subdir)
+          _ -> GithubConfig (T.unpack str) Nothing
 
     homepage :: Maybe String
     homepage = case packageConfigHomepage of
       Just Nothing -> Nothing
       _ -> join packageConfigHomepage <|> fromGithub
       where
-        fromGithub = ((++ "#readme") <$> github)
+        fromGithub = (<> "#readme") . githubConfigUrl <$> github
```
Comment (sol) [inline src/Config.hs]: As we stick with `String` I would prefer to not change this to `<>` (as `<>` is not available in older versions of base).

File: test/RunSpec.hs [hunk 9]
```diff
@@ -150,3 +138,18 @@ spec = do
           , "  ghc-options: -Wall -Werror"
           , "  default-language: Haskell2010"
           ]
+
```
Comment (sol) [inline test/RunSpec.hs]: The following specs should be group under `describe "renderSourceRepository"`

File: test/RunSpec.hs [hunk 10]
```diff
@@ -150,3 +138,18 @@ spec = do
...
+    it "renders source-repository without subdir correctly" $ do
+      renderSourceRepository (GithubConfig "https://github.com/hspec/hspec" Nothing)
+        `shouldBe` unlines [ "source-repository head"
+                           , "  type: git"
+                           , "  location: https://github.com/hspec/hspec"
+                           ]
```
Comment (sol) [inline test/RunSpec.hs]: ^ please don't align.  Use e.g. the following instead:

```
`shouldBe` unlines [
    "source-repository head"
  , "  type: git"
  , "  location: https://github.com/hspec/hspec"
  ] 
```

Comment (sol) [inline test/RunSpec.hs]: My reason is that if you change this code to e.g. `(T.pack . unlines) [...` later on, you have to realign the whole list.

**Comment by sol** [comment]
@Fuuzetsu Yes, I think your approach regarding `String` and `Text` is good.  Other then my comments, LGTM.

**Comment by Fuuzetsu** [comment]
See latest commit.

**Pull request closed without merging**